reactor-core/src/main/java/reactor/core/publisher/FluxZip.java

```diff
@@ -653,9 +684,45 @@ void cancelAll() {
 			}
 		}
 
-		void drain() {
+		void discardAll(int m) {
+			final Context context = actual.currentContext();
+			final Object[] values = current;
+			Operators.onDiscardMultiple(Arrays.asList(values), context);
+			Arrays.fill(values, null);
+
+			for (;;) {
+				for (ZipInner<T> s : subscribers) {
+					final Queue<T> queue = s.queue;
+					final int sourceMode = s.sourceMode;
+
+					if (queue != null) {
+						if (sourceMode == ASYNC) {
+							// delegates discarding to the queue holder to ensure there is no racing on draining from the SpScQueue
```

Comment:
As of the fusion contract, the call for the `queue.clear` means - termination of the process. In case of fusion - queue comes from the upstream, so the upstream is a creator / owner of the queue, so it and only it can have full control of when values are offered and drained. See https://github.com/reactor/reactor-core/issues/2636 for more information

Reply:
thanks, that's what I was asking about.